aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,466 @@
+
+  if (dim >= sparse_dim) {
+    if (LogSoftMax) {
+      auto new_values = log_softmax_cpu(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
+    } else {
+      auto new_values = softmax_cpu(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
+    }
+    return;
+  }
+
+  auto nnz = values.size(0);
+  scalar_t* values_data_base = values.data_ptr<scalar_t>();
+  iscalar_t* indices_data_base = indices.data_ptr<iscalar_t>();
+  scalar_t* out_values_data_base = out_values.data_ptr<scalar_t>();
+  /* Compute pool indices */
+  auto sizes = input.sizes().vec();
+
+  std::vector<iscalar_t> offsets = get_offsets<iscalar_t>(indices, sizes, dim);
+  iscalar_t mx_p;
+  std::vector<iscalar_t> pool = get_dense_offsets<iscalar_t>(mx_p, offsets);
+
+  /* Compute mx */
+  std::vector<int64_t> mx_sizes(sizes.begin() + sparse_dim - 1, sizes.end());
+  mx_sizes[0] = mx_p;
+  at::Tensor mx = at::empty(mx_sizes, values.options());
+  scalar_t* mx_data_base = mx.data_ptr<scalar_t>();
+  auto nvalues = get_nvalues(sizes, sparse_dim);
+
+  auto ninf = -std::numeric_limits<scalar_t>::infinity();
+  for (int64_t j=0; j < nvalues * mx_p; j++) {
+    mx_data_base[j] = ninf;
+  }
+
+  for (int64_t i=0; i < nnz; i++) {
+    auto p = pool[i];
+    scalar_t* mx_data = mx_data_base + p * nvalues;
+    scalar_t* values_data = values_data_base + i * nvalues;
+    for (int64_t j=0; j < nvalues; j++) {
+      mx_data[j] = std::max(mx_data[j], values_data[j]);
+    }
+  }
+
+  /* apply exp to (v - mx) and sum the results */
+  at::Tensor exp_sums = at::zeros_like(mx);
+  scalar_t* exp_sums_data_base = exp_sums.data_ptr<scalar_t>();
+  for (int64_t i=0; i < nnz; i++) {
+    auto p = pool[i];
+    scalar_t* mx_data = mx_data_base + p * nvalues;
+    scalar_t* values_data = values_data_base + i * nvalues;
+    scalar_t* out_values_data = out_values_data_base + i * nvalues;
+    scalar_t* exp_sums_data = exp_sums_data_base + p * nvalues;
+    for (int64_t j=0; j < nvalues; j++) {
+      auto v = std::exp(values_data[j] - mx_data[j]);
+      if (!LogSoftMax) {
+        out_values_data[j] = v;
+      }
+      exp_sums_data[j] += v;
+    }
+  }
+
+  for (int64_t j=0; j < nvalues * mx_p; j++) {
+    if (LogSoftMax) {
+      mx_data_base[j] += std::log(exp_sums_data_base[j]);
+    } else {
+      exp_sums_data_base[j] = 1.0 / exp_sums_data_base[j];
+    }
+  }
+
+  /* normalize with the sum of exponents */
+  for (int64_t i=0; i < nnz; i++) {
+    auto p = pool[i];
+    scalar_t* values_data = values_data_base + i * nvalues;
+    scalar_t* out_values_data = out_values_data_base + i * nvalues;
+    scalar_t* exp_sums_data = exp_sums_data_base + p * nvalues;
+    scalar_t* mx_data = mx_data_base + p * nvalues;
+    for (int64_t j=0; j < nvalues; j++) {
+      if (LogSoftMax) {
+        out_values_data[j] = values_data[j] - mx_data[j];
+      } else {
+        out_values_data[j] *= exp_sums_data[j];
+      }
+    }
+  }
```

Comment:
I would use TensorIterator here.

Reply:
You could also use accessors to eliminate raw pointer exposures.

Reply:
I have found no evidence that using TensorIterator would perform better here.
Even when assuming `nvalues << nnz`, the need to use the `pool` mapping makes the current loops most effective for the considered test cases (see `test/test_sparse.py::test_softmax`).

Reply:
Could you please show how you used TensorIterator here?

Reply:
And I mean the loop over `nnz`, not `nvalues`.

Reply:
@nikitaved , `cpu_sparse_coo_softmax` has five for-loops, only the last loop is independent in writes for different `i in range(nnz)`, and the first and the 4rd are independent in writes for different `i in range(npools)`. In most cases, `npools << nnz`.

Reply:
Here are cumulative timings for the five loops when running `test_softmax`:
```
loop 1 took     366590 ns
loop 2 took     329048 ns
loop 3 took    2075724 ns
loop 4 took     774612 ns
loop 5 took     349260 ns
```
As you can see, the 3rd loop dominates, most likely due to the `exp` calls.
The 4th loop has also `log` call but this loop over `range(npools)` is short compared to the loop over `range(nnz)`.

Reply:
It is possible to refactor the loop over `range(nnz)` to independent sub-loops.
Consider the 3rd loop (for simplicity, using scalar dense part):
```
for (int64_t i=0; i < nnz; i++) {
  auto p = pool[i];
  auto v = exp(values[i] - mx[p]);
  out_values[i] = v;
  exp_sums[p] += v;
}
```
that can be replaced with
```
for (int64_t p=0; p < npools; p++) {
  auto mx_p = mx[p];
  scalar_t acc = 0;
  for (int64_t i=0; i < nnz; i++) {
    if (pool[i] == p) {
      auto v = exp(values[i] - mx_p);
      out_values[i] = v;
      acc += v;
    }
  }
  exp_sums[p] = acc;
}
```
so that the loops over `p` are now independent and could be run in parallel.

Can you suggest how to set up a TensorIterator for the above?

Reply:
Notice that with the above loop-refactoring all five loops could be joined under one loop over `p in range(npools)` which makes the approach very interesting.

Reply:
All loops are now restructured to use independent pools of indices (=> increase of performance about 25%) and using parallel_for (=> additional increase of performance about 7-8%).